traffic_ops/traffic_ops_golang/deliveryservice/request/requests.go

```diff
@@ -20,451 +22,843 @@ package request
-s.username AS assignee,
-r.assignee_id,
-r.author_id,
-r.change_type,
-r.created_at,
-r.id,
-r.last_edited_by_id,
-r.last_updated,
-r.deliveryservice,
-r.status,
-r.deliveryservice->>'xmlId' as xml_id
+	downgraded := make([]tc.DeliveryServiceRequestV15, 0, len(dsrs))
+	for _, dsr := range dsrs {
+		downgraded = append(downgraded, dsr.Downgrade())
+	}
 
-FROM deliveryservice_request r
-JOIN tm_user a ON r.author_id = a.id
-LEFT OUTER JOIN tm_user s ON r.assignee_id = s.id
-LEFT OUTER JOIN tm_user e ON r.last_edited_by_id = e.id
-`
-	return query
+	api.WriteResp(w, r, downgraded)
 }
 
-// IsTenantAuthorized implements the Tenantable interface to ensure the user is authorized on the deliveryservice tenant
-func (req TODeliveryServiceRequest) IsTenantAuthorized(user *auth.CurrentUser) (bool, error) {
+// isTenantAuthorized ensures the user is authorized on the DSR's Requested
+// and/or Original Delivery Service's Tenant, as appropriate to the change type.
+func isTenantAuthorized(dsr tc.DeliveryServiceRequestV30, inf *api.APIInfo) (bool, error) {
+	if dsr.Requested != nil && (dsr.ChangeType == tc.DSRChangeTypeUpdate || dsr.ChangeType == tc.DSRChangeTypeCreate) {
+		if dsr.Requested.TenantID == nil {
+			log.Debugf("requested.tenantID is nil")
+			return false, errors.New("requested.tenantID is nil")
+		}
+		ok, err := tenant.IsResourceAuthorizedToUserTx(*dsr.Requested.TenantID, inf.User, inf.Tx.Tx)
+		if err != nil {
+			err = fmt.Errorf("requested: %v", err)
+		}
+		if !ok || err != nil {
+			return ok, err
+		}
+	}
 
-	ds := req.DeliveryService
-	if ds == nil {
-		// No deliveryservice applied yet -- wide open
+	ds := dsr.Original
+	if ds == nil || dsr.ChangeType == tc.DSRChangeTypeCreate {
+		// No deliveryservice applied yet or change type doesn't require an original
 		return true, nil
 	}
+
 	if ds.TenantID == nil {
-		log.Debugf("tenantID is nil")
-		return false, errors.New("tenantID is nil")
+		log.Debugf("original.tenantID is nil")
+		return false, errors.New("original.tenantID is nil")
+	}
+	ok, err := tenant.IsResourceAuthorizedToUserTx(*ds.TenantID, inf.User, inf.Tx.Tx)
+	if err != nil {
+		err = fmt.Errorf("original: %v", err)
 	}
-	return tenant.IsResourceAuthorizedToUserTx(*ds.TenantID, user, req.APIInfo().Tx.Tx)
+	return ok, err
 }
 
-// Update implements the tc.Updater interface.
-//all implementations of Updater should use transactions and return the proper errorType
-//ParsePQUniqueConstraintError is used to determine if a request with conflicting values exists
-//if so, it will return an errorType of DataConflict and the type should be appended to the
-//generic error message returned
-func (req *TODeliveryServiceRequest) Update() (error, error, int) {
-	if req.ID == nil {
-		return errors.New("missing id"), nil, http.StatusBadRequest
+// Warning: this assumes inf isn't nil, and neither is dsr, inf.Tx or inf.User or inf.Tx.Tx.
+func insert(dsr *tc.DeliveryServiceRequestV30, inf *api.APIInfo) (int, error, error) {
```

Comment:
Just a convention. `golint` says that errors should always be the last values returned from a function - which is a convention followed throughout the standard library.

We have a golangci-lint Dockerfile in the repository, but we don't as a project actually enforce any go linting or vetting beyond `go fmt`, so if you want me to change it for consistency, I will.

Reply:
IMO `(error, error, int)` should be exempt from that rule because an HTTP status code is effectively an error code. I can see why you would choose the order that you did though, so go with what you think is best.

It would be nice to enforce linting at some point. In order for `golangci-lint` to ignore that particular `golant` rule, we should exclude it in `issues.exclude-rules` in our `.golangci.yml`.

Reply:
> It would be nice to enforce linting at some point

It might, but I doubt we ever will. Go fmt is already enough to annoy people. They'll be very mad if we add enforcements of vetting or linting. It's not a hill I'm willing to die on, is what I'm saying.

Reply:
Oh, I agree we shouldn't add it all at once. We could disable all the rules that fail so that the master branch passes and only slowly enabling more rules 1 by 1 as we improve the codebase.

Reply:
Well anyway, I'd like to leave it as-is for now. It passes linting and that pleases me.

Also, though, I think it'd be nice to do something like 
```go
type Errors {
    Code             int
    SystemError error
    UserError     error
}

func (e Errors) Occurred() bool {
    return e.SystemError != nil || e.UserError != nil
}
```

Because I, for one, always have to look up the function I'm calling to figure out which error is for the system and which is for the user, and I think this would make things a lot easier, and solve the linting problems at the same time.

Reply:
From @rob05c in #3349:

> The `tc.ApiErrorType` is a legacy artifact that needs removed. TO Golang has standardized around the `userErr, sysErr, errCode` idiom, which is more powerful and flexible, and easier to work with (no type casting). The few remaining functions in TOGo using `tc.ApiErrorType` need changed to use the `userErr, sysErr, errCode` idiom, and the symbol removed.

A struct seems like an improvement over the original iota enum, and yeah it would satisfy the `error should be the last type when returning multiple items` golint rule. However, since, as #3349 mentions, `error, error, int` is the current standard,
* `error, error, int` should be used here. Project standard > project-agnostic standard
* We should reach some degree of consensus about whether we want to make that switch. Let's open an issue for it in the meantime, to document it as a possibility.

Reply:
Well nothing not discussed on the mailing list is a project standard.

Reply:
Ha, fair